File: py/algorithm/auto_encoder_model.py
import numpy as np
import py.algorithm.neural_network_model as nn

class AutoEncoder():

    def train(self, samples):
        dimension = samples[0].shape[0]
        compressed_dimension = int(np.ceil(dimension / 2))
        if compressed_dimension <= 0:
            compressed_dimension = 1
        self.neural_network_model = nn.NeuralNetworkModel(np.array([dimension, compressed_dimension, dimension]))
        self.neural_network_model.train(samples, samples)

    def encode(self, x):
        self.neural_network_model.forward(x)
        output_of_all_layers = self.neural_network_model.output_mat
        output = output_of_all_layers[1]
        return output

File: src/main/java/jm/app/algorithm/AutoEncoderModel.java
package jm.app.algorithm;

import jm.app.algebra.Matrix;

public class AutoEncoderModel {

    private FastNeuralNetworkModel neuralNetworkModel;

    private Integer compressedDimension;

    public AutoEncoderModel() {
    }

    public AutoEncoderModel(int compressedDimension) {
        this.compressedDimension = compressedDimension;
    }

    public void train(Matrix[] samples) {
        int dimension = samples[0].getRowNum();
        if (null == compressedDimension) {
            compressedDimension = dimension / 2 > 0 ? dimension / 2 : 1;
        }
        neuralNetworkModel = new FastNeuralNetworkModel(dimension, compressedDimension, dimension);
        neuralNetworkModel.train(samples, samples);
    }

    public Matrix encode(Matrix x) {
        neuralNetworkModel.forward(x);
        Matrix[] outputOfAllLayers = neuralNetworkModel.getOutputMat();
        Matrix output = outputOfAllLayers[1];
        return output;
    }

    public Integer getCompressedDimension() {
        return compressedDimension;
    }

    public void setCompressedDimension(Integer compressedDimension) {
        this.compressedDimension = compressedDimension;
    }
}
